modules/lru_cache.py

```diff
@@ -0,0 +1,94 @@
+from threading import RLock
+
+
+class _Node:
+    def __init__(self, key):
+        self.key = key
+        self.next = None
+        self.prev = None
+
+
+class LruCache:
+    def __init__(self):
+        self._cache = {}
+        self._head = _Node(0)
+        self._tail = _Node(0)
+        self._head.next = self._tail
+        self._tail.prev = self._head
+        self._lock = RLock()
+
+    def pop(self):
+        """
+        remove a key from cache and return the removed key
+        Returns:
+
+        """
+        with self._lock:
+            if len(self._cache) > 0:
+                node = self._head.next
+                self._link_remove(node)
+                del self._cache[node.key]
+                return node.key
+            return None
+
+    def touch(self, key):
+        """
+        touche a key.
+        if this key is not in cache, it will be added to link tail.
+        Args:
+            key:
+
+        Returns:
+
+        """
+        with self._lock:
+            if key not in self._cache:
+                node = _Node(key)
+                self._cache[node.key] = node
+                self._link_add(node)
+                return
+            current = self._head.next
```

Comment:
Since you already use `self._cache = {}`, which is a dictionary, why don't you just
```
node = self._cache[key]
```
instead of iterating through the link list?